components/org.wso2.carbon.identity.recovery/src/test/java/org/wso2/carbon/identity/recovery/internal/service/impl/username/UsernameRecoveryManagerImplTest.java

```diff
@@ -0,0 +1,413 @@
+        when(mockUserAccountRecoveryManager.getUserRecoveryData(recoveryCode, RecoverySteps.SEND_RECOVERY_INFORMATION))
+                .thenReturn(mockUserRecoveryData);
+        when(mockUserRecoveryData.getRemainingSetIds()).thenReturn("SMS,SMS");
+        when(Utils.getRecoveryConfigs(anyString(), anyString())).thenReturn(TRUE);
+        when(Utils.resolveEventName(anyString())).thenReturn("TRIGGER_SMS_NOTIFICATION_LOCAL");
+        User mockUser = new User();
+        mockUser.setUserName("KD123");
+        mockUser.setTenantDomain(TENANT_DOMAIN);
+        when(mockUserRecoveryData.getUser()).thenReturn(mockUser);
+        UsernameRecoverDTO result = usernameRecoveryManager.notify(recoveryCode, "2", TENANT_DOMAIN, properties);
+        assertEquals(result.getCode(), "UNR-02001");
+        assertEquals(result.getMessage(), "Username recovery information sent via user preferred notification channel.");
+    }
+
+    /**
+     * Test to validate callback URL.
+     *
+     * @throws IdentityRecoveryException if an error occurs during validation.
+     */
+    @Test
+    public void testCallbackURLValidation() throws IdentityRecoveryException {
+
+        mockIdentityEventService();
+        String callbackURL = "http://localhost:8080";
+        String recoveryCode = UUID.randomUUID().toString();
+        Map<String, String> properties = new HashMap<>();
+        properties.put("useLegacyAPI", TRUE);
+        properties.put(IdentityRecoveryConstants.CALLBACK, callbackURL);
+        mockedJDBCRecoveryDataStore.when(JDBCRecoveryDataStore::getInstance).thenReturn(mockUserRecoveryDataStore);
+        mockedRecoveryManagerStatic.when(UserAccountRecoveryManager::getInstance).thenReturn(mockUserAccountRecoveryManager);
+        when(mockUserAccountRecoveryManager.getUserRecoveryData(recoveryCode, RecoverySteps.SEND_RECOVERY_INFORMATION))
+                .thenReturn(mockUserRecoveryData);
+        when(mockUserRecoveryData.getRemainingSetIds()).thenReturn("SMS,SMS");
+        when(Utils.getRecoveryConfigs(anyString(), anyString())).thenReturn(TRUE);
+        when(Utils.resolveEventName(anyString())).thenReturn("TRIGGER_SMS_NOTIFICATION_LOCAL");
+        User mockUser = new User();
+        mockUser.setUserName("KD123");
+        mockUser.setTenantDomain(TENANT_DOMAIN);
+        when(mockUserRecoveryData.getUser()).thenReturn(mockUser);
+        when(Utils.handleClientException(
+                IdentityRecoveryConstants.ErrorMessages.ERROR_CODE_CALLBACK_URL_NOT_VALID, callbackURL))
+                .thenReturn(new IdentityRecoveryClientException(null));
+        usernameRecoveryManager.notify(recoveryCode, "2", TENANT_DOMAIN, properties);
+    }
+
+    /**
+     * Test to get callback URL with exception.
+     *
+     * @throws IdentityRecoveryException if an error occurs during retrieval.
+     */
+    @Test(expectedExceptions = NullPointerException.class)
```

Comment:
From where the NPE is thrown. Ideally we should avoid NPE. The test should cover the failure.

Reply:
Check other test cases also written for NPE